crates/subspace-core-primitives/src/lib.rs

```diff
@@ -117,6 +116,120 @@ pub const REWARD_SIGNATURE_LENGTH: usize = 64;
 const VRF_OUTPUT_LENGTH: usize = 32;
 const VRF_PROOF_LENGTH: usize = 64;
 
+/// Representation of a single BLS12-381 scalar value.
+#[derive(Debug, Default, Copy, Clone, Eq, PartialEq)]
+pub struct Scalar(Fr);
+
+impl TryFrom<&[u8]> for Scalar {
+    type Error = ();
+
+    /// Number of bytes must be exactly [`Self::SAFE_BYTES`] bytes.
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        if value.len() != Self::SAFE_BYTES {
+            return Err(());
+        }
+
+        let mut bigint_bytes = [0u64; 4];
+        // NOTE: `bigint_bytes` has one more byte than `bytes`, the last byte will always be zero.
+        bigint_bytes
+            .iter_mut()
+            .zip(value.chunks(mem::size_of::<u64>()))
+            .for_each(|(output, input): (&mut u64, &[u8])| {
+                let mut tmp = 0u64.to_le_bytes();
+
+                tmp.iter_mut().zip(input).for_each(|(output, input)| {
+                    *output = *input;
+                });
+
+                *output = u64::from_le_bytes(tmp);
+            });
+
+        Ok(Scalar(
+            Fr::from_repr(BigInteger256::new(bigint_bytes))
+                .expect("Always smaller than modulus, always 31 bytes; qed"),
+        ))
+    }
```

Comment:
Maybe you can use here `TryFrom<&[u8]> for &[u8; SAFE_BYTES]` and after that run `.into()` on `&[u8; SAFE_BYTES]`?

Reply:
I like the suggestion, applied